pyfolio/round_trips.py

```diff
@@ -13,168 +13,253 @@
+    portfolio_value : pd.Series (optional)
+        Portfolio value (all net assets including cash) over time.
+
     Returns
     -------
-    transactions_split : pd.DataFrame
-        Prices and amounts of executed trades. Trades that cause
-        total position amount to cross zero are divided.
+    round_trips : pd.DataFrame
+        DataFrame with one row per round trip.  The returns column
+        contains returns in respect to the portfolio value while
+        rt_returns are the returns in regards to the invested capital
+        into that partiulcar round-trip.
     """
 
-    trans_split = []
+    transactions = _groupby_consecutive(transactions)
+    roundtrips = []
 
     for sym, trans_sym in transactions.groupby('symbol'):
         trans_sym = trans_sym.sort_index()
-
-        while True:
-            cum_amount = trans_sym.amount.cumsum()
-            # find the indicies where position amount crosses zero
-            sign_flip = np.where(np.abs(np.diff(np.sign(cum_amount))) == 2)[0]
-
-            if len(sign_flip) == 0:
-                break  # all sign flips are converted
-
-            sign_flip = sign_flip[0] + 2
-
-            txn = trans_sym.iloc[:sign_flip]
-
-            left_over_txn_amount = txn.amount.sum()
-            assert left_over_txn_amount != 0
-
-            split_txn_1 = txn.iloc[[-1]].copy()
-            split_txn_2 = txn.iloc[[-1]].copy()
-
-            split_txn_1['amount'] -= left_over_txn_amount
-            split_txn_2['amount'] = left_over_txn_amount
-
-            # Delay 2nd trade by a second to avoid overlapping indices
-            split_txn_2.index += pd.Timedelta(seconds=1)
-
-            assert split_txn_1.amount.iloc[0] + \
-                split_txn_2.amount.iloc[0] == txn.iloc[-1].amount
-            assert trans_sym.iloc[:sign_flip - 1].amount.sum() + \
-                split_txn_1.amount.iloc[0] == 0
-
-            # Recreate transactions so far with split transaction
-            trans_sym = pd.concat([trans_sym.iloc[:sign_flip - 1],
-                                   split_txn_1,
-                                   split_txn_2,
-                                   trans_sym.iloc[sign_flip:]])
-
-        assert np.all(np.abs(np.diff(np.sign(trans_sym.amount.cumsum()))) != 2)
-        trans_split.append(trans_sym)
-
-    transactions_split = pd.concat(trans_split)
-
-    return transactions_split
+        price_stack = deque()
+        dt_stack = deque()
+        for dt, t in trans_sym.iterrows():
+            if t.price < 0:
+                warnings.warn('Negative price detected, ignoring for'
+                              'round-trip.')
+                continue
+            signed_price = t.price * np.sign(t.amount)
+            abs_amount = int(abs(t.amount))
+            indiv_prices = [signed_price] * abs_amount
+            if (len(price_stack) == 0) or \
+               (copysign(1, price_stack[-1]) == copysign(1, t.amount)):
+                price_stack.extend(indiv_prices)
```

Comment:
If `signed_price * abs_amount` is always a float, you could make `indiv_prices = signed_price * abs_amount` and do `price_stack.append(indiv_prices)` instead of `price_stack.extend(indiv_prices)`